### PR TITLE
Scenario improvements

### PR DESCRIPTION
**Pull request by calderinisystemorph**
Content: minor changes to Yield Curve comparer and importer, Scenario validation for any type of importer

### CODE REVIEW

File: ifrs17/Constants/Validations.ipynb [hunk 1]
```diff
@@ -137,6 +137,8 @@
                 "\n    (Error.PortfolioGricNotFound          , 2) => $\"Portfolio {s[0]} assigned to Group of Reinsurance Contract {s[1]} does not exist.\",",
                 "\n    (Error.InvalidAmountTypeEstimateType  , 2) => $\"Invalid combination of EstimateType {s[0]} and AmountType {s[1]}.\",",
                 "\n    (Error.MultipleTechnicalMarginOpening , 1) => $\"Multiple opening for techincal margin are not allowed for DataNode {s[0]}.\",",
+                "\n    (Error.DimensionNotFound              , 2) => $\"Property {0} has unknown value {1}.\",",
```
Comment (dcolleoni) [inline ifrs17/Constants/Validations.ipynb:140]: Instead of "Property" I would say "Column".

File: ifrs17/Import/Importers.ipynb [hunk 2]
```diff
@@ -367,9 +367,11 @@
         {
             "cell_type": "code",
             "source": [
-                "public static void ValidateArgsForPeriod(this ImportArgs args) {",
+                "public async static void ValidateArgsForPeriod(this ImportArgs args, IDataSource targetDataSource) {",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:370]: rename ValidateArgsForPeriod => ValidateArgsForPeriodAsync

File: ifrs17/Import/Importers.ipynb [hunk 3]
```diff
@@ -862,25 +865,25 @@
...
                 "\n        ",
                 "\n        // Avoid starting the computation if no best estimate cash flow has ever been imported ",
-                "\n        if(args.Scenario == null) {",
-                "\n            await options.TargetDataSource.Partition.SetAsync<PartitionByReportingNodeAndPeriod>(null);",
-                "\n            if(!(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any()) continue;",
-                "\n        }",
+                "\n        if(!(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == defaultPartition).Take(1).ToArrayAsync()).Any() &&",
+                "\n           !(await options.TargetDataSource.Query<IfrsVariable>().Where(x => x.Partition == defaultPartition).Take(1).ToArrayAsync()).Any()) continue;",
+                "\n        ",
+                "\n        // Only nominals corresponding to the target data nodes are added to the workspace",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:882]: rename (and comment) : instead of nominal just say rawVariables. 
(Now Nominals is a computed EconomicBasis)

**Comment by dcolleoni** [comment]
Change value of Scenarios.EnableScenario

File: ifrs17/Import/Importers.ipynb [hunk 4]
```diff
@@ -1243,9 +1251,9 @@
                 "\n        await options.TargetDataSource.Partition.SetAsync<PartitionByReportingNodeAndPeriod>(null);",
                 "\n        if(ApplicationMessage.HasErrors()) return Activity.Finish().Merge(log);",
                 "\n        ",
-                "\n        // Avoid starting the computation if no best estimate cash flow and actuals has ever been imported ",
-                "\n        if(args.Scenario == null && !(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any() &&",
-                "\n                                    !(await options.TargetDataSource.Query<IfrsVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any()) continue;",
+                "\n        // Avoid starting the computation if no best estimate cash flow or actuals has ever been imported ",
+                "\n        if(!(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == defaultPartition).Take(1).ToArrayAsync()).Any() ||",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:1255]: checking ifrsVariable is enough. 
In the end when RawVariables are created, also IfrsVariables are.

File: ifrs17/Import/Importers.ipynb [hunk 5]
```diff
@@ -833,63 +836,68 @@
...
-                "\n            Month = primaryArgs.Month, ",
-                "\n            Scenario = primaryArgs.Scenario,",
-                "\n            Values = datarow.Table.Columns.Where(c => c.ColumnName.StartsWith(nameof(YieldCurve.Values))).OrderBy(c => c.ColumnName.Length).ThenBy(c => c.ColumnName)",
-                "\n                            .Select(x => datarow.Field<string>(x.ColumnName).CheckStringForExponentialAndConvertToDouble()).ToArray(),",
-                "\n            Name = hasNameColumn ? datarow.Field<string>(nameof(YieldCurve.Name)) : default(string)",
+                "\n    var log = await Import.FromDataSet(dataSet).WithType<YieldCurve>(",
+                "\n        (dataset, datarow) => {",
+                "\n            var values = datarow.Table.Columns.Where(c => c.ColumnName.StartsWith(nameof(YieldCurve.Values))).OrderBy(c => c.ColumnName.Length).ThenBy(c => c.ColumnName)",
+                "\n                            .Select(x => datarow.Field<string>(x.ColumnName).CheckStringForExponentialAndConvertToDouble()).ToArray().Prune();",
+                "\n            if (values.Length == 0 ) return null;",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:852]: !values.Any() instead of checking Length

**Comment by calderinisystemorph** [comment]
https://github.com/Systemorph/IFRS17CalculationEngine/pull/279#issuecomment-1520843635
done

**Pull request merged**